### ISSUE TITLE
Release versions with webtorrent support?

### ISSUE THREAD

**Issue by petterreinholdtsen** [arvidn/libtorrent#7283]
Hi.  I tried and failed to figure out which library releases include webtorrent support.  Can you help point me to a version with such support available?

I had a look at #4123, but it did not realy make this clear.

**Comment by SilentBot1** [comment]
Hey @petterreinholdtsen, 

Only the master branch, built with the `-DWebtorrent=on` flag set has WebTorrent support. From my understanding there are no existing releases with this enabled by default and you must compile this yourself.

Hope this helps.

**Comment by petterreinholdtsen** [comment]
[Brad Marsden]
> Hey @petterreinholdtsen, 
>
> Only the master branch, built with the `-DWebtorrent=on` flag set has
> WebTorrent support. From my understanding there are no existing
> releases with this enabled by default and you must compile this
> yourself.
>
> Hope this helps.

It is useful knowledge, at least.  Is there any hope to have a tagged
release soon that I can ask to have packaged in Debian, for use by the
general public?

-- 
Happy hacking
Petter Reinholdtsen

**Comment by tuxayo** [comment]
You can try LibreTorrent on Android https://f-droid.org/packages/org.proninyaroslav.libretorrent/
It's client that is compiled with the support enabled.

If you do testing, be careful to disable any webseed, non-wss tracker, DHT, PEX and LSD (local seed discovery) be sure to have a webtorrent peer and not a regular one. If it works, check on the peer list the user agent and connection of the peer and tell us what you got :)

**Comment by Donokami** [comment]
Any update on when will we get WebTorrent support on libtorrent by default?

**Comment by vrubleg** [comment]
This is a really bad way to resolve tickets.

**Comment by petterreinholdtsen** [comment]
I similar request is in https://github.com/arvidn/libtorrent/issues/7333, where a user ask for webtorrent support in the default python module.

**Comment by bertung** [comment]
Is there a reason why web torrent support ıs left out of release builds?

**Comment by kingomarnajjar** [comment]
WE WANT WEBTORRENT BY DEFAULT! (pretty pretty please :)
@arvidn is a sex machine for creating and maintaining Libtorrent. It is a GOD made library. To change the world of torrenting, moving to webtorrent enabled by default completely changes the internet and the world. 
I don't think its possible to exaggerate the impact of the world doing browser peer to peer file sharing. 
It's one of those sounds boring but insanely impactful when you look back at it

Please be the king of torrents and make this god like feature a reality for the masses by default.

**Comment by OvercookedBeef** [comment]
I think the reason it still isn't enabled by default is arvidn is against the protocol. He might believe it is bad for the torrent economy because web seeders aren't really going to stay around to seed. But webtorrent is still extremely important for two use cases:

1. **Previews**: Torrent sites often show image previews of the content. The problem with this is that they don't want to host the previews themselves, likely for legal reasons, so it is required to upload them to a 3rd-party image host. But when you go to look at multi-year old torrents, the images are likely either taken down or the host is gone. WebTorrent could solve this because it means distributed hosting of the images, and since previews are tiny, it won't have a big cost on the seeders.
2. **Extremely easy web-based P2P**: I can imagine many services would enjoy a non-centralized file storage. For example PeerTube is a P2P video hosting software (although they switched away from WebTorrent). WebTorrent makes this super easy to do and allows people to help increase speeds (although you would likely need an HTTP source for low latency, it still helps with big files, like videos). Of course there are other WebTorrent clients like the official JS one that services could use, but libtorrent is much more mature, scalable (I would assume), and is widely used which would lower the barrier of entry for people who want to help seed web content.

**Important**: This is an **assumption** of his reasoning so don't take it as fact. AFAIK he hasn't stated dislike of the protocol anywhere. But if this is the case, I hope this changes your opinion.

**Comment by arvidn** [comment]
it's primarily that the attack surface of libtorrent is significantly larger with webtorrents enabled. It's especially disappointing to be hacked by a feature you're not using.

**Comment by arvidn** [comment]
surely any client that wants to support webtorrents would enable it. Why is the default so significant?

**Comment by petterreinholdtsen** [comment]
[Arvid Norberg]
> surely any client that wants to support webtorrents would enable
> it. Why is the default so significant?

Are you talking about it being build it by default and enabled at
runtime, or not being built in by default and a rebuild is needed to
enable it?

Only enabling it at runtime work for clients sharing the same shared
library.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by fineless71** [comment]
> surely any client that wants to support webtorrents would enable it. Why is the default so significant?

My understanding was that WebTorrent was added to the `master` branch, not any of the release branches, like `RC_2_0`. I do not have a good understanding of libtorrent's structure so correct me if I'm wrong.

It seems many pieces of software depend on the releases or release branch:
https://github.com/qbittorrent/qBittorrent/issues/17974#issuecomment-1302616587
https://github.com/transmission/transmission/issues/47#issuecomment-1272237178
https://github.com/arvidn/libtorrent/pull/4123#issuecomment-1863985198

But if you think the implementation has security issues, that obviously takes priority. Have you received any reports about this? I couldn't find anything in this repo about security issues with WebTorrent. Have you experienced this yourself? 

> Only enabling it at runtime work for clients sharing the same shared
library.

I also agree with this. Clients won't want to require custom builds of libtorrent that they can't control. I don't think clients typically static link libtorrent.

**Comment by Ghost-chu** [comment]
> surely any client that wants to support webtorrents would enable it. Why is the default so significant?

As https://github.com/qbittorrent/qBittorrent/issues/17974#issuecomment-1302616587 says, qBittorrent is just waiting for libtorrent to announce it GA.
I guess if the feature isn't enabled by default, we might not get enough user feedback to push it to become GA.

The Libre client, as well as BiglyBT (via the WebTorrents plugin) are proof that the standard seems to work and can have a realistically positive effect.
libtorrent also apparently has the code to join the party, but it's disabled by default, so I see clients like qBittorrent still waiting in the wings.

We have a deadlock - it's not the default option because few users use it - and so no more users will use it and push to make it a default option.

The potential of WebTorrents is huge, as loading a JS on a browser and connecting to a BitTorrent network is far less difficult than instructing users on how to properly install and configure their BT client.
Thanks to browsers, any user can easily run a WebTorrents client on PCs, Laptops, and cell phones. the BitTorrent community will gain a lot of Peers and everyone will benefit from it.
Whether they are traditional BT clients or browser-based WebTorrents clients, once they are able to connect to each other, they will become part of the network, expanding the size of BitTorrent's network.

**Comment by bertung** [comment]
WebTorrent support provides a significant boost for those who do not want to store their data on big tech cloud platforms. Instead, they can share it with peers, making the data accessible directly in web browsers. This is a major breakthrough. It goes beyond simple usage statistics—it has the potential to disrupt the entire ecosystem. For this reason, qbittorrent should have WebTorrent support included in its releases.

**Comment by WrayOfSunshine** [comment]
Looking forward to when webtorrent hits libtorrent so qbittorrent can use it.

**Comment by vrubleg** [comment]
I'm afraid that people with WebTorrent won't seed for long time. It will be used mostly to download something and immediately stop seeding.

**Comment by petterreinholdtsen** [comment]
[Evgeny Vrublevsky]
> I'm afraid that people with WebTorrent won't seed for long time. It
> will be used mostly to download something and immediately stop
> seeding.

What is the basis for your fear, any why do you fear it?

I suspect most webtorrent users will be watching what they fetch, and
would seed it for the duration of the content they downloaded, like
video and audio from archive.org.  I do not know if you consider it a
long time or not, but it seem to me it would help all others watching
the same content for that duration. :)

-- 
Happy hacking
Petter Reinholdtsen

**Comment by Ghost-chu** [comment]
> I'm afraid that WebTorrent will be used mostly to download something without seeding for long enough time to keep torrents alive.

What are you worried about? You know, the users who don't want to do the seeding, and even if they use a BT client, they simply close the software immediately after downloading.

A 1.75GB 40-minute long 1080p video can be downloaded in less than a minute using a BT client, and then that user is immediately offline.
WebTorrents' streaming keeps users online for at least 20 minutes, and you only need ~15Mbps of upstream bandwidth to have a positive effect on the network (increasing the share rate to 1.0 or higher).
If a user watches a 40-minute video in its entirety, then 5Mbps is enough.

We would still need the legacy BT client as a fallback to keep the seed active for long periods of time, and this issue is discussing that.

**Comment by bertung** [comment]
> I'm afraid that WebTorrent will be used mostly to download something without seeding for long enough time to keep torrents alive.

I am no really into sharing videos. I am taking about web sites and other kinds of data that can be seeded such as compact LLM models. This will make sharing permanent, at least for the data which is not video.

**Comment by InevitableAmbrosia** [comment]
I want to use WebTorrent for my public domain ebook/audiobook website propagate.info. I have it working but BiglyBT was too buggy to seed 3,000 torrents. I switched to libtorrent (which I read can seed 10,000 torrents) and I hope you guys come around to implementing it in the future.

**Comment by QuixThe2nd** [comment]
I second the runtime flag. I develop many web-based P2P projects. The lack of any real WebTorrent adoption has really stifled my progress. Considering every major Bittorrent client is waiting on Libtorrent to include WebTorrent in releases, allowing webtorrent to be toggled in runtime instead of at build would be very beneficial.

**Comment by kingomarnajjar** [comment]
> I second the runtime flag. I develop many web-based P2P projects. The lack of any real WebTorrent adoption has really stifled my progress. Considering every major Bittorrent client is waiting on Libtorrent to include WebTorrent in releases, allowing webtorrent to be toggled in runtime instead of at build would be very beneficial.
> 
> This really seems like a chicken and egg problem, of no user adoption because no clients support it, and no clients support it because there is no user adoption. Adding the option to enable WebTorrent without needing a seperate build would solve this. This only gets more critical as time goes on considering the trend of % of user time spent on the web increasing. I hope @arvidn would consider this. WebTorrent adoption would seriously revive the torrent scene and libtorrent adoption is the primary hurdle we need to overcome to make widespread WebTorrent adoption a reality.
> 
> > it's primarily that the attack surface of libtorrent is significantly larger with webtorrents enabled. It's especially disappointing to be hacked by a feature you're not using.
> 
> Addressing this, I completely understand the importance of addressing the larger surface area of attack. But this will never truely be addressed without actual usage. As for "a feature you're not using", I believe this is disingenuous as no one is using WebTorrent because its not enabled. Again, this is a chicken and egg problem, that can only be solved by actually enabling the feature.
> 
> I and many other devs are willing to contribute our time towards solving any issues with WebTorrent implementation if you point us in the right direction. I really want to see WebTorrent adoption become a reality, and this just won't ever happen unless we bite the bullet.

I second this. Chicken goes baaak baaak
Eggs are expensive now unless you just get a chicken.
Soo many devs will incorporate this, it'll be a massive revival of torrenting on a scale I don't think even you or any cyberpunk can imagine.

@arvidn stop being a bitch. You are a god for libtorrent now add webtorrent at runtime and ride that golden wave of status and sex machineness. I will literally get you a hot gamer girl or 3 if you do this. 

All of us watching this thread have hundreds of devs working with us and will clean up anything required.
As god of libtorrent @arvidn we go at your command.

**Comment by kingomarnajjar** [comment]
![image](https://github.com/user-attachments/assets/b33a1182-7670-4b3b-8f7b-b9254a9c9799)
@arvidn o captain my captain!

**Comment by petterreinholdtsen** [comment]
Given the current progress on this issue, I suspect a more productive way forward is to fork the project under a new name, and provide it as an alternative with webtorrent enabled.  Similar to how sqlcipher is a fork of sqlite with encryption enabled.

**Comment by QuixThe2nd** [comment]
> Given the current progress on this issue, I suspect a more productive way forward is to fork the project under a new name, and provide it as an alternative with webtorrent enabled.  Similar to how sqlcipher is a fork of sqlite with encryption enabled.

This doesn't seem like a productive way forward. The reason so many people are asking for it to be included in a release build is because libtorrent is a core dependency in majority of BitTorrent clients. I don't see clients like qBitTorrent changing forks considering they've specified that they are waiting for it to become GA in libtorrent.

The only concern @arvidn has mentioned regarding enabling webtorrent's WebRTC transport in a release is a larger surface area of attack for those not using it, which is mute if it is enabled as a runtime flag instead. I'm not sure what his real motive behind delaying this for 5 years is, but it would be nice if he could articulate his reasoning. As this issue is a major factor in the death of BitTorrent and he's been quiet in the face of dozens of people explaining why it's so important.

**Comment by thalieht** [comment]
The delay has nothing to do with webtorrent: https://github.com/arvidn/libtorrent/pull/7377#discussion_r1162537632

**Comment by kingomarnajjar** [comment]
> The delay has nothing to do with webtorrent: https://github.com/arvidn/libtorrent/pull/7377#discussion_r1162537632

What are you talking about? That issue you linked is merged, makes no sense. Clarify.
@thalieht

**Comment by thalieht** [comment]
>What are you talking about? That issue you linked is merged, makes no sense. Clarify.

I linked to a comment in the issue. A pread based, disk I/O backend is in the works but it's still not ready.
>I'd like to release 2.1.0 as soon as the disk cache for pread-disk-io is done

2.1.0 will be based on the master branch which contains webtorrent support in case it's not clear.

**Comment by QuixThe2nd** [comment]
> > What are you talking about? That issue you linked is merged, makes no sense. Clarify.
> 
> I linked to a comment in the issue. A pread based, disk I/O backend is in the works but it's still not ready.
> 
> > I'd like to release 2.1.0 as soon as the disk cache for pread-disk-io is done
> 
> 2.1.0 will be based on the master branch which contains webtorrent support in case it's not clear.

@thalieht WebTorrent was PRed to the master branch 5 years ago, prior to 2.0 even being released. Webtorrent has been excluded despite new versions being released since. How is that issue related to webtorrent?

**Comment by thalieht** [comment]
>WebTorrent was PRed to the master branch 5 years ago, prior to 2.0 even being released.

RC_2_0 branch was created before webtorrent was merged into master.

**Comment by QuixThe2nd** [comment]
> > WebTorrent was PRed to the master branch 5 years ago, prior to 2.0 even being released.
> 
> RC_2_0 branch was created before webtorrent was merged into master.

Keyword being **RC**. How many versions came out since then? How is what you said related?

**Comment by thalieht** [comment]
>Keyword being **RC**. How many versions came out since then? How is what you said related?

And it stayed that way... latest 2.0 and 1.2 versions are in their respective  `RC_2_0` and `RC_1_2` branches.

**Comment by QuixThe2nd** [comment]
> > Keyword being **RC**. How many versions came out since then? How is what you said related?
> 
> And it stayed that way... latest 2.0 and 1.2 versions are in their respective `RC_2_0` and `RC_1_2` branches.

How come versions after RC2 didnt include webtorrent then. I dont get how your original comment is relataed.

**Comment by thalieht** [comment]
>How come versions after RC2 didnt include webtorrent then.

Because the master branch is the bleeding edge branch. In this repository development is done in the stable branches and these are then merged to master. Some things like ABI breaking changes are only in master. I don't know if anything has ever been backported from master to a stable branch but webtorrent is not one of them.
If you want further proof search for `webtorrent` in the changelog in `RC_2_0` branch https://github.com/arvidn/libtorrent/blob/RC_2_0/ChangeLog
and then look at the changelog in the master branch: 
https://github.com/arvidn/libtorrent/blob/c48cdadcdb453b89c1f97bf7dd7485ec8bd376e9/ChangeLog#L18

> I dont get how your original comment is related.

Because i erroneously assumed that everyone here is familiar with the workflow of this repository.

**Comment by QuixThe2nd** [comment]
Its been 5 years since webtorrent was merged in master. How long are we expected to wait before RC2.1 is released? The workflow should be changed if it takes over half a decade for an accepted PR to be included in a release, and @arvidn should state that it will be included in 2.1 instead of letting dozens of people assume that he doesn't want it included at all. The lack of communication & clarity is the reason why there are so many comments on both this issue and #7333.

**Comment by Ghost-chu** [comment]
How many 5 years are there in a human life?
The fact that it can only be enabled by compile-time means determines how many people are not involved in its testing. If no one tests it, this concern will never go away.

Runtime enablement is the way to really put it to the test in practice. Honestly, has anyone really audited a section of code in 5 years that doesn't even have a feature enabled by default, I'm sure the code just lying there quietly isn't going to have any positive effect on either security or functionality.

**Comment by QuixThe2nd** [comment]
Would appreciate some actual communication from @arvidn to break this deadlock and to actually understand your intentions. None of us enjoy hassling you over this and truely appreciate your work, but this is a critical feature, and we have been waiting for over half a decade. WebTorrent came out over a decade ago. The only communication we've received from you in the 5 years people have been waiting is concern over security, which can be used to prevent ANY feature from being included in a release and is mute if we go down the runtime toggle path. I have asked both claude and chatgpt for their thoughts and they both sided with the community about including webtorrent in a release.

"If I had to take a side, I would align with those advocating for WebTorrent inclusion as a runtime-toggleable feature in release builds" https://claude.ai/share/785311ae-de6b-4836-9b05-b92fa4c0872c

"from a practical and forward-looking perspective, the community is more correct" https://chatgpt.com/share/681afbf9-d688-8011-ae4c-419982324d89

**Comment by sneakers-the-rat** [comment]
Seeing the discussion above re: webtorrent and hit and runs, and seeing the discussion mostly focused on peertube and video - we are also working on a bittorrent-backed web archive in collaboration with some of the folks who make the tech used by a certain famous web archive site, and we need webtorrent in libtorrent so that we can mix longterm seedboxes/dedicated seeders with ephemeral browser-based peers. without webtorrent in libtorrent, we can't make a bittorrent-powered web archive, since all the seedboxes are running ootb clients like qbittorrent and transmission and not a custom-compiled libtorrent.

**Comment by QuixThe2nd** [comment]
It's a very unfortunate situation considering BitTorrent usage would be much higher if this repository didn't exist. The current maintainers are stifling innovation and there's nothing anyone can do about it other than pester them...

**Comment by Ghost-chu** [comment]
> It's a very unfortunate situation considering BitTorrent usage would be much higher if this repository didn't exist. The current maintainers are stifling innovation and there's nothing anyone can do about it other than pester them...

This is completely wrong, and the moral abduction of developers through this means is nothing short of disgraceful.
The libtorrent repository plays a crucial role in the BitTorrent ecosystem, it is the most complete BitTorrent implementation and is still under active development today.
Without libtorrent, we wouldn't have Deluge, we wouldn't have qBittorrent, and many other familiar BitTorrent programs, and in fact the repository is good enough and powerful enough for CPP developers.
When I talk about a feature of libtorrent not being accepted by developers, I hope I'm making a strong enough case to convince developers to do so, rather than "to hell with this unruly repository that's disappeared from the world" while simultaneously splitting up and destroying the hard-won ecosystem and community of BitTorrent.

**Comment by Aenigma** [comment]
I, too, would love for webtorrent to be more widely available as I've started playing with the idea of a web based publication platform that makes archival transparent that can last beyond traditional hosting. That being said...

@QuixThe2nd I think you're really out of line right now. You're telling the maintainer of libtorrent that it would be better if this repository didn't exist for BitTorrent because he won't smash the release button for you.

The reality is that only a minority of developers care about webtorrent. It looks to me that webtorrent development has stalled and there's relatively little interest in it these days. I think it's telling that it seems like outside of libtorrent, the only library code interacting with webtorrent clients is using the official JavaScript webtorrent library. Of course, there's definitely a catch-22 problem here, but the solution isn't insulting the maintainers, it's seeing how you can participate in the webtorrent ecosystem. I've seen issues and PRs that @sneakers-the-rat here has participated in webtorrent repos just today and it's been inspiring to me about the health of the webtorrent ecosystem.

Maybe we can look into distributing precompiled builds of libtorrent with webtorrent support and seeing if there would be issues building qbittorrent or deluge with webtorrent support. I would love if we could get something using libtorrent to seed webtorrent within docker or something while we're waiting for official support.

**Comment by QuixThe2nd** [comment]
> Insults are never a good idea at any time or place for any reason. Please stop doing this and leave constructive comments.

I never insulted the maintainer. I expressed the true statement that this repository is hindering bittorrent growth.

**Comment by QuixThe2nd** [comment]
> it's always nice when a big new feature is finally mainlined and it feels like a win, rather than the end of a long struggle, and it might be good to try and cultivate that :)

@sneakers-the-rat I completely agree, but its been 5 years. At what point will the maintainer start reciprocating? Its literally been half a decade and nothing has happened. What else are we meant to do?

**Comment by kingomarnajjar** [comment]
How much money do you want directly to include it? @arvidn 

We'll set up a pool as this issue is already like 2 years old and webtorrent implementation has been 5 years

**Comment by vrubleg** [comment]
I believe that supporting WebTorrent will boost creation of services that abuse BitTorrent ecosysten without adding any value back. It will just motivate and support people who download and immediately leave without seeding. Current default motivates people to install a proper client that is designed to seed for long time, at least most of popular clients are like this.

If you need any proofs, just check how many WebTorrent seeds are in any existing WebTorrent download. People don't keep browser tabs open just to seed a little longer. It's just a joke that BitTorrent needs WebTorrent, it's quite the opposite.

If anybody needs to seed WebTorrent for their website, just compile your own libtorrent and install a client that uses it on your seed boxes. But don't expect that you can abuse existing infrastructure.

**Comment by Ghost-chu** [comment]
> I believe that supporting WebTorrent will boost creation of services that abuse BitTorrent ecosysten without adding any value back. It will just motivate and support people who download and immediately leave without seeding. Current default motivates people to install a proper client that is designed to seed for long time, at least most of popular clients are like this.
> 
> If you need any proofs, just check how many WebTorrent seeds are in any existing WebTorrent download. People don't keep browser tabs open just to seed a little longer. It's just a joke that BitTorrent needs WebTorrent, it's quite the opposite.
> 
> If anybody needs to seed WebTorrent for their website, just compile your own libtorrent and install a client that uses it on your seed boxes. But don't expect that you can abuse existing infrastructure.

While you are watching a video, WebTorrent shares the data while getting it from other people.
Theoretically, as long as your upload speed transfer is greater than the transfer needed for the playback speed, the share rate should be 1.0 or higher.

So when a video is watched by tens of thousands of people at the same time, this is where WebTorrent will work its magic.

It's been a long time since this Issue opened and the world has changed a bit, due to the popularity of short videos, videos have become shorter in length and users are staying in tabs for shorter periods of time, which makes me have some concerns.

I'm not sure if this can be remedied with Service Worker, such as keeping the last 100 and video do kind of as long as the user opens any page.
Extensions would also be a solution, I've seen some extensions that reside in my taskbar and perform background tasks even after the browser is closed.

**Comment by vrubleg** [comment]
If WebTorrent seeders could add any value, the WebTorrent ecosystem wouldn't struggle.

**Comment by Ghost-chu** [comment]
> > So when a video is watched by tens of thousands of people at the same time, this is where WebTorrent will work its magic.
> 
> If "thousands" of WebTorrent seeders could add any value, the WebTorrent ecosystem wouldn't struggle.
> 
> It will harm BitTorrent ecosystem.

The concerns are real and it's impossible to predict where WT will really go until it really catches on.

**Comment by QuixThe2nd** [comment]
> > So when a video is watched by tens of thousands of people at the same time, this is where WebTorrent will work its magic.
> 
> If "thousands" of WebTorrent seeders could add any value, the WebTorrent ecosystem wouldn't struggle.
> 
> It will harm BitTorrent ecosystem.

If WebTorrent support is enabled through a runtime flag, this is a non-issue and all sides are happy.

**Comment by kingomarnajjar** [comment]
The fear is ridiculous. Invention requires we move forward and adapt with the times.
The internet hope by cypherpunks was to be decentralised so no one, no gov could shut down the people's will for the freedom of information.
This is a battle we are losing. Significantly.
I want to win again to quote the orange man.

This is because of complacency and fear. Nothing more, nothing less.

Players like cloudflare being amazing and centralising every bloody thing.

I'm sick of the monopolisation on information whether that be movies, apps, historical archives or anything else.

Information wants to be free, webtorrent was created to be that next gen layer and hasn't been adopted by its elders because of ego it seems.
It's time to unite and forge a future where our great players aren't scared to step up to the plate of the present and write a great history.

The people watching this issue belong to strong power bases.

The movie scene will adopt this as it's a no brainer for streaming content as the upload rate will always be above 1.0.
It doesn't mean having a torrent client in browser just means all the streaming websites don't have to hide their servers constantly.

The iOS scene will adopt this because they believe in it. 

The archive scene will adopt this to keep history accurate. 

That's 3 obvious scenes and we haven't even started to make it popular.

Add to runtime please as a step closer to decentralisation.
And as others have said this doesn't make it a default; just included in runtime so everyone can simply turn it on with a flag.

If Steve Jobs was in our scene, he'd bully us all into it as it's the right move forward.

Not fear and complacency. 

We need vision and guts.

**Comment by sneakers-the-rat** [comment]
> I believe that supporting WebTorrent will boost creation of services that abuse BitTorrent ecosysten without adding any value back.

If this is part of the reason why webtorrent support is held back from mainline, i think this is worth addressing. 

What we're trying to do is build a hybrid p2p system that consists of *both* long-term seeds *and* temporary, ephemeral peers. In our model (and in how bittorrent has worked, always, forever) this is a *good thing.* Having low barrier to entry does mean that it is possible for people to hit and run, but it also means that more people can participate. In the long history of bittorrent, the latter force has proven over and over again to win out - sure it is true that not everyone seeds to 150%, but a lot of people do choose to voluntarily share their resources and enable the whole system to thrive.

The idea that we should choose to resist some hypothetical impact on swarms at large by resisting it at an implementation level is, in my opinion, the wrong venue and the wrong goal. The argument that "the webtorrent people need to prove their worth by cultivating a resilient seeder atmosphere" gets the problem backwards - webtorrent is a *complementary* technology that allows *more* people to get involved with bittorrent passively. In our case, if it was possible for the 100 people that were viewing a website at a time to drop in, seed to a few other ephemeral viewers, have the whole swarm backed by a few dozen long-term seeds, that would be a *huge* win for bittorrent and for the whole p2p ecosystem. 

Keeping the scene closed to preserve it the way it is is how the scene dies. Resisting new protocols is how the scene dies. Bittorrent has has a very long history *precisely because* of its adaptatability to new technologies and new adversaries. Extending to the web and allowing an *enormous range* of new modalities of using bittorrent as we know it can only be a good thing - if you want to ban webtorrent peers, the option should be open to refuse connections for webtorrent peers, but please don't let that be an argument that holds the rest of the bittorrent world that wants to use it in different contexts and in different ways back.

**Comment by kingomarnajjar** [comment]
**69th comment**

**Comment by sneakers-the-rat** [comment]
at the risk of double commenting, i would also add, would not a bittorrent-powered archive.org be sick as hell

**Comment by InevitableAmbrosia** [comment]
I just wanted to give my input here. I'm not a big time coder or anything but I made this website, propagate.info. My whole point was to allow anyone to download public domain educational media using WebTorrent in the Browser. I can seed from my Desktop if there is a WebTorrent-supported client. I think others will also be able to WebTorrent seed with qBitTorrent or Deluge. People want to support free educational media distribution. In my use-case, most PDFs are under 30mb, so as long as there are like 2-3 WebTorrent seeds from supporters of public domain information, there shouldn't be a problem with Browser hit and runs. In fact, Browser hit and runs are EXACTLY what we want, because it allows people who aren't so tech savvy to have free and seamless access to public domain ebooks and audiobooks. One reason for my website is, as we saw when Internet Archive went down 6 months ago, it is good for Civil Society when the public literally owns their own information. Anyway, I got a bit disillusioned with the project after working on it for about 10 years. My ideal was Gazelle web 3.0 with WebTorrent clients. I may not renew the domain when it expires, but if libtorrent comes out with WebTorrent support, I will renew the domain and continue the project. I just write this so you can see what is seriously at stake, and to assuage some concerns about Browser hit and runs. Thank you for listening.

**Comment by Shrinks99** [comment]
From @sneakers-the-rat:
> I would also add, would not a bittorrent-powered archive.org be sick as hell
> 
> like the wayback machine part. the internet archive part.

I'll take this opportunity to add that Webrecorder & some other folks I'm working with are currently working on adding Webtorrent support to [ReplayWeb.page](https://webrecorder.net/replaywebpage/) — Webrecorder's web archive viewer.  Right now archives hosted on sites like [SciOp](https://sciop.net/) could become instantly viewable in the browser.  This is something that is being actively developed but will only become _cool_ when the pool of peers that support Webtorrent is reliably large.  Having this enabled by default in libtorrent would make a meaningful difference in making this a reality!

Decentralized web archive hosting seems like a worthy use case to add to the pile :)

**Comment by sneakers-the-rat** [comment]
Yes I didn't want to name drop y'all too directly since I wasn't sure if that work was public, but for everyone else's benefit, webrecorder/replayweb.page was what I was referring to above, and I'm here as core dev of sciop - we have a big use case in hand that specifically requires both webtorrent and libtorrent with wt support. The pieces are all in place except for webtorrent in downstream clients like qbt.

**Comment by petterreinholdtsen** [comment]
Did anyone so far try to write a patch to make webtorrent support in
libtorrent a runtime setting?

I notice the code is protected by #ifdef RTC_ENABLE_WEBSOCKET, and all
of the code is located in the libdatachannel git submodule, spread
across 41 files.  I guess a new runtime setting would have to be similar
to the other boolean options declared in
include/libtorrent/settings_pack.hpp.  Do not know the code base and
lack the spare time to write such patch myself, but perhaps someone else
got the spare time needed?

-- 
Happy hacking
Petter Reinholdtsen

**Comment by tuxayo** [comment]
@kingomarnajjar WTF stop putting so much pressure, please chill.

There is at least this article that would help see the situation more accurately: https://boehs.org/node/bountysource

Bonus:
https://tylercipriani.com/blog/2023/01/15/entitlement-burnout-and-toxicity/

https://mikemcquaid.com/entitlement-in-open-source/
https://devtechinsights.com/open-source-maintainer-burnout/

**Comment by tuxayo** [comment]
Back directly on topic: if anyone can help by testing a web torrent with the python client or some other way. That would be appreciated. Because **there might be big blocker bug**.
On LibreTorrent (Android client that uses libtorrent) I get segfaults only when seeding a webtorrent, tons of segfaults. It might be the libtorrent4j layer. But it might very very well libtorrent itself. libtorrent4j is "just" a wrapper and segfaults are a C/C++ thing. It's not certain, but that's not a good sign.
So here we would have an obvious reason it's not production ready.

example torrent: https://webtorrent.io/torrents/sintel.torrent
bonus: After adding the torrent, leave only WSS trackers and disable PeX and DHT. So your traffic will likely only be webtorrent peers.

Here is my failed attempt at building the python client: https://github.com/arvidn/libtorrent/discussions/7906

On LibreTorrent, I had **7 crashes per day on average.**[1] So if you manage to test libtorrent directly, you should know for sure after a few days of seed time if that big issue is on libtorrent itself.

[1] It depend on the traffic to https://webtorrent.io/. Because that's what it's seeding to.

**Comment by QuixThe2nd** [comment]
It's been almost 3 years since this issue was created and @arvidn has averaged 1 comment a year. I feel like we're talking to a wall. It would be nice if you could at least acknowledge that you've read all the core points made.

No one likes the fact that this issue is needed, but honestly, what other option do we have than forking (which no one wants to do)? It feels like you're dying on a hill you don't need to die on. Like it or not, this repo is critical infrastructure for almost all of torrenting, and we have no option but to beg. It's a very unfortunate situation for everyone involved, it'd be nice if we could get some sort of communication from the only guy who has the authority to decide how this proceeds.

**Comment by sneakers-the-rat** [comment]
acknowledgement or at least some list of things that need to get done in order for webtorrent to get defaulted. we've been applying for funding for bt client work, it would be great if we could use that funding to help with libtorrent, but if it isn't moving/isn't interested in webtorrent we'll have to commit to some of the newer rust-based projects instead. 

i'm not sure which of these are actually the blocker:
- personal/community feelings about the validity of webtorrent (the HnR question, see [above](https://github.com/arvidn/libtorrent/issues/7283#issuecomment-2911287354)) 
- security concerns about the threat surface of webtorrent
- need for a runtime switch for webtorrent
- webtorrent isn't actually working (segfaulting comment above)
- maintainer exhaustion about hearing the word "webtorrent"

but if we got some feedback about what needs to get done here, we could hopefully resolve the deadlock.

**Comment by i990049** [comment]
If anyone wants to try the hybrid protocol of webtorrent and bittorrent, you can use the following software
https://webtorrent.io/desktop/
https://www.softpedia.com/get/Internet/Streaming/Popcorn-Time-Desktop.shtml
https://www.wiztorrent.com/

**Comment by kingomarwashere** [comment]
https://github.com/kingomarwashere/libtorrent
⚡ POLITICAL PROTEST FORK ⚡
The war has begun. 
@arvidn RIP master.

**Comment by pzmarzly** [comment]
[#8532: enable webtorrent by default](https://github.com/arvidn/libtorrent/pull/8532) merged

**Comment by sneakers-the-rat** [comment]
Great news! Thanks arvid!!!

**Comment by i990049** [comment]
Can the RC4 encryption protocol be used with WebRTC? My ISP has already restricted BitTorrent downloads, so I have to enable RC4 to use a BT client.



---Original---
From: "Jonny ***@***.***&gt;
Date: Mon, Jul 6, 2026 07:00 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [arvidn/libtorrent] Release versions with webtorrent support? (Issue #7283)


sneakers-the-rat left a comment (arvidn/libtorrent#7283)
 
Great news! Thanks arvid!!!
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!

**Comment by SilentBot1** [comment]
@i990049 as WebRTC is an already encrypted data channel, you don't need both WebRTC and protocol encryption enabled at the same time, but the outcome should still be the same.

**Comment by i990049** [comment]
Is WebRTC only available in bt 2.0 and hybrid torrents? many private tracker don't want to support bt 2.0 and hybrid torrents. Is it possible let libtorrent 1.x support WebRTC? so private tracker can use WebRTC to transfer bt 1.0 torrents.
&nbsp;


i990049
***@***.***



        



         原始邮件
         
       
发件人：Brad Marsden ***@***.***&gt;
发件时间：2026年7月10日 23:59
收件人：arvidn/libtorrent ***@***.***&gt;
抄送：i990049 ***@***.***&gt;, Mention ***@***.***&gt;
主题：Re: [arvidn/libtorrent] Release versions with webtorrent support? (Issue #7283)



SilentBot1 left a comment (arvidn/libtorrent#7283)

@i990049&nbsp;as WebRTC is an already encrypted data channel, you don't need both WebRTC and protocol encryption enabled at the same time, but the outcome should still be the same.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS&nbsp;and Android. Download it today!

**Comment by SilentBot1** [comment]
@i990049 libtorrent RC_2_1 enables WebRTC by default, but the library implements BitTorrent v1, v2 and supports hybrid torrents.

WebRTC is a connection type, similar to i2p, uTP and TCP, with the BitTorrent v1 and v2 protocol being able to go over this connection.

WebRTC connections require WebTorrent/websocket trackers due to the need to signal WebRTC offers and answers, which are different than the standard UDP and HTTP trackers, so private trackers would need to support this tracker type to support WebRTC peers.
